### ISSUE TITLE
E2E Dry-run helper

### ISSUE THREAD

**Issue by johnSchnake** [vmware-tanzu/sonobuoy#1524]
Lots of users over the years have always wanted to know what tests exist, what tags exist, etc etc.

This changes with each k8s release slightly and I had in mind to make a helper that allowed the following:

```
# Get tests; makes it much faster if you just want to try dry-run stuff
$ sonobuoy e2e tests --e2e-focus=foo --e2e-skip=bar --kubernetes-version=v1.21.3
< lists all the tests that would match >

# Just reads the tags from the tests
$ sonobuoy e2e tags <takes input from stdin>
< lists all the tags in that test, unique/sorted >

# Then you can combine them to check tags of certain tests if desired
$ sonobuoy e2e tests --e2e-focus=foo --e2e-skip=bar --kubernetes-version=v1.21.3 | sonobuoy e2e tags
```

And you can see the difference between two releases as well, or see how many tags conformance does/doesn't cover. Maybe even add a flag to not just grab the tags/tests but to show counts as well.

I wanted to make this a web page with a nice UI but this is very easy to implement and someone can always wrap it up with some nicer UI if helpful.

**Comment by johnSchnake** [comment]
@NikhilSharmaWe do you have interest in doing this? If so I can help with getting all the core data that sonobuoy could use and it would provide you some experience with the CLI code.

**Comment by NikhilSharmaWe** [comment]
@johnSchnake of course I would like to work on it.

**Comment by johnSchnake** [comment]
OK, so the setup I'm envisioning for this first pass (we just want to get the feature out there and see if it resonates) is this:
 - I'll get lists of k8s-version + test list for each release and put them into github
 - You add a command into sonobuoy CLI `sonobuoy e2e tests` with the flags mentioned above.
 - I'd want another flag as well to specify a test list manually (either a file or a URL). This gives the users a way to work around any shortcoming with our data (either it is wrong at some point or missing a certain version)

So the logic of the commands is pretty straight forwards; the `e2e tests` command will just get a list of tests and apply the regexp for focus to filter the tests, then apply the skip filter to further reduce the list.

The e2e tags command will take a string like `foo [bar] [baz]` and return all the strings within brackets. In this case, `bar` and `baz`. Iterate over the tests and make the values sorted/unique.

I'd add another flag on the `e2e tags` command: `--mode=[list|listcount|all]` so instead of outputting:
```
# with list mode (default)
bar
baz

# with list count
bar: 2
baz: 14

# with all
bar
bar
baz
baz
baz
...
```

That way you can see how heavily weighed a particular tag is and can make meaningful diffs between two different versions/regexp settings.

Does all that make sense? Let me know what kind of timeframe you think you'll need for this; happy to do a few review sessions with you too to help move you forward when stuck.

**Comment by NikhilSharmaWe** [comment]
I have understood what basically we are trying to accomplish here. Can you give me some code pointers from where I can start adding the new commands `sonobuoy e2e tests` and then `sonobuoy e2e tags <tests name>`.

**Comment by NikhilSharmaWe** [comment]
@johnSchnake I have understood what basically we are trying to accomplish here. Can you give me some code pointers from where I can start adding the new commands sonobuoy e2e tests and then sonobuoy e2e tags <tests name>. In which files the code should be added for the purpose. I am working for the first time on adding a new command in the CLI, so I am little bit confused here but I will give my best.

**Comment by johnSchnake** [comment]
You can see an example of adding a command here: https://github.com/vmware-tanzu/sonobuoy/blob/main/cmd/sonobuoy/app/status.go#L71-L93

That creates a function that returns a cobra.Command object.

Then you need to add it to the root command appropriately: https://github.com/vmware-tanzu/sonobuoy/blob/main/cmd/sonobuoy/app/root.go#L57

I'm still tossing the ideas around in my head about what command names make sense and will fit into sonobuoy CLI structure the best. As described above, there are actually 3 commands: `sonobuoy e2e` (which does nothing, just has subcommands), `sonobuoy e2e tests`, and `sonobuoy e2e tags`. So you'll create all 3 commands, add the 2 subcommands to the `e2e` command, and add the `e2e` command to the root command.

**Comment by NikhilSharmaWe** [comment]
@johnSchnake Am I thinking reasonable here ?
First we need to add a function like NewE2ECommand() in `sonobuoy/app/root.go` similar to [this](https://github.com/vmware-tanzu/sonobuoy/blob/main/cmd/sonobuoy/app/root.go#L31) which does nothing but is the root command for other 2 commands. What should be the logic inside that function, just a map like in this case ? 

After that we will make a `sonobuoy/app/e2e.go` file in which we will write functions for our subcommands.
If I am correct till now, then how we will connect the root command and subcommands.

**Comment by johnSchnake** [comment]
Look [here](https://github.com/vmware-tanzu/sonobuoy/blob/main/cmd/sonobuoy/app/root.go#L45-L51); you can see that we create the `sonobuoy gen` command and then get other commands and add to `gen`, then add `gen` onto the root.

[This](https://github.com/vmware-tanzu/sonobuoy/blob/main/cmd/sonobuoy/app/images.go#L55-L91) is another example (slightly different organization) where we create all the subcommands and attach them when returning the `images` command (so the root.go code doesn't have to even know about it).

If a cobra.Command doesn't have any logic to perform, thats ok: it just needs either `Run`(or `RunE`) OR children commands. Thats the way I recall it at least.

**Comment by johnSchnake** [comment]
Fixed via #1587 and #1588
